### ISSUE TITLE
Fast refresh error - can't access property "trim", t is null

### ISSUE THREAD

**Issue by frehner** [pmmmwh/react-refresh-webpack-plugin#216]
## Issue

In [this example repo](https://github.com/frehner/react-fast-refresh-externals), I get the following error 100% of the time:

`Uncaught (in promise) TypeError: can't access property "trim", t is null`

in debugging, it comes from this function 

https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/89ccca3a528a68f06f777daef09a28a5614e5525/sockets/utils/getCurrentScriptSource.js#L5

which does various things to try and figure out the current script source. However, in that example repo, `document.currentScript` is null AND `currentScript.getAttribute('src');` is null - which then causes issues down the line: specifically here (where the error is eventually thrown)

https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/89ccca3a528a68f06f777daef09a28a5614e5525/sockets/utils/getSocketUrlParts.js#L22

## Fix proposal

With that information, I propose the following change to the `getCurrentScriptSource` (with the caveat that I am not a webpack plugin developer, so I don't know the ins-and-outs of webpack plugins and this may not be the correct solution. However, I have tested this proposal and it appears to fix everything in that example repo)

```js
getCurrentScriptSource() {
  return __webpack_public_path__
}
```

which, again, fixes the issue for that example repo... and potentially fixes any issues for anyone using a [dynamic public path](https://webpack.js.org/guides/public-path/#on-the-fly) as well?

If you think that's the correct way to go, then I can make a PR for it. 

Thanks for this project, and I'm looking forward to your thoughts.

**Comment by frehner** [comment]
Hello, just wondering if there's any additional information you need from me for this bug? Thanks.

**Comment by pmmmwh** [comment]
> Hello, just wondering if there's any additional information you need from me for this bug? Thanks.

Hi, I'm sorry but I don't have time to deal with this until this weekend. I think I know what needs to happen for a fix and hopefully that works so I can probably push out a new version sometime next week.

**Comment by frehner** [comment]
> > Hello, just wondering if there's any additional information you need from me for this bug? Thanks.
> 
> Hi, I'm sorry but I don't have time to deal with this until this weekend. I think I know what needs to happen for a fix and hopefully that works so I can probably push out a new version sometime next week.

no worries. thank you for taking the time to look at it. :)

**Comment by frehner** [comment]
if you have an idea of what needs to be done, I would also be willing to try and implement it if you can point me in the right direction. :)

**Comment by frehner** [comment]
I was able to find this code block that html-webpack-plugin uses, and I think it may apply here as well. I'll test it out tomorrow 

https://github.com/jantimon/html-webpack-plugin/blob/6bff0c139ae8497483cd756ba5bb8386a96a5c88/index.js#L535-L563

**Comment by frehner** [comment]
I've found that if you use `__webpack_public_path__` as mentioned in the first issue, and change the order of the injected entries in this file https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/lib/utils/injectRefreshEntry.js

then it correctly can pick up when an app dynamically changes the public path. So it appears to me that this would probably be the best solution. 

your thoughts?

**Comment by pmmmwh** [comment]
> I've found that if you use `__webpack_public_path__` as mentioned in the first comment, and change the order of the injected entries in this file [https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/lib/utils/injectRefreshEntry.js](https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/lib/utils/injectRefreshEntry.js?rgh-link-date=2020-10-24T16%3A44%3A33Z)
> 
> then it correctly can pick up when an app dynamically changes the public path. So it appears to me that this would probably be the best solution.

What do you mean by changing the order of the injected entries?

Also now I'm wondering why `webpack-dev-server` does not have a similar issue since I did reference what they do quite a bit while coming up with this fix for public paths ...

**Comment by frehner** [comment]
>What do you mean by changing the order of the injected entries?

a problem happens when an app sets public path dynamically at runtime - if you look at this line, 

https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/lib/utils/injectRefreshEntry.js#L87

you're see it's injecting the socketEntries before the originalEntry. but that means that the socket code in this plugin will execute _before_ the app code dynamically sets the public path, which will cause the socket connection to look at the wrong spot. 

However, if you put originalEntry first, and then socketEntries, then it allows the app code to execute first which means that it will set the dynamic path in time for this plugin's socket code to correctly pick it up.

Does that make sense? It's kind of confusing haha. 

I can even put together a PR if you would like

**Comment by frehner** [comment]
I've been reading webpack-dev-server's source to figure out what they do, and I think I have a good idea:

When they are creating the socketUrl, they first check for `resourceQuery` 

https://github.com/webpack/webpack-dev-server/blob/4ab1f21bc85cc1695255c739160ad00dc14375f1/client-src/default/utils/createSocketUrl.js#L8-L26

which is a special global that webpack adds. [docs for resourceQuery](https://webpack.js.org/api/module-variables/#__resourcequery-webpack-specific)

that comes from here, where the clientEntry has the resourceQuery added to it

https://github.com/webpack/webpack-dev-server/blob/4ab1f21bc85cc1695255c739160ad00dc14375f1/lib/utils/addEntries.js#L30-L41

------

with that knowledge, perhaps it would be best to follow that pattern as well?

**Comment by patrikholcak** [comment]
I just encountered the exact same problem, though my circumstances are bit more specific. I am running an inlined script inside of Figma (all extensions run inside of an iframe using data uri).

So the `document.currentScript` for the inlined script doesn’t have a src and the url parsing fails. I can’t really change that. I was able to fix the error by doing exactly what @frehner is suggesting - parsing the `resourceQuery` first and commenting out the other code in `getSocketUrlParts`. It takes precedence anyway, so I didn’t really understand why the implementation is reversed and the `resourceQuery` is only called after everything is constructed from `currentScript` src

Furthermore, due to the way figma plugins work (mainly the data uri, and requirement for https) would it be possible to also add `sockProtocol` to `overlay` options? This is literally the last thing that’s holding me back, I’m willing to help in any way and submit PR with the fix.

This approach doesn’t introduce any breaking changes as the `resourceQuery` is already used and takes precedence

**Comment by frehner** [comment]
@pmmmwh Ok, I'm feeling better about the PR I just created - #241 

It doesn't require changing any tests, and mainly just adds the resourceQuery so that `getSocketUrlParts` can use that first, and then fall back to the current method if resourceQuery isn't there. 

I was able to test it in my example repo above and it worked great - reloads worked, and the error overlay worked as well. However, I still have it as a draft PR just because I'm not as familiar with this stuff as you are. 

Would love to know your thoughts on if this is getting close to the real solution or not. Thanks!
